Framework/API/inc/MantidAPI/Algorithm.tcc

```diff
@@ -0,0 +1,142 @@
+  declareProperty(
+      Kernel::make_unique<WorkspaceProperty<T>>(
+          propertyName, "", Kernel::Direction::Input, optional, lock),
+      doc);
+
+  declareProperty(Kernel::make_unique<IndexTypeProperty>(
+      propertyName + "IndexType", allowedIndexTypes));
+
+  auto *wsProp =
+      dynamic_cast<WorkspaceProperty<T> *>(getPointerToProperty(propertyName));
+  auto *indexTypeProp = dynamic_cast<IndexTypeProperty *>(
+      getPointerToProperty(propertyName + "IndexType"));
+
+  declareProperty(Kernel::make_unique<IndexProperty>(propertyName + "IndexSet",
+                                                     *wsProp, *indexTypeProp));
+
+  m_reservedList.push_back(propertyName);
+  m_reservedList.push_back(propertyName + "IndexType");
+  m_reservedList.push_back(propertyName + "IndexSet");
+}
+
+template <typename T1, typename T2, typename WsType>
+void Algorithm::doSetIndexProperty(const std::string &name, const T1 &wksp,
+                                   IndexType type, const T2 &list) {
+  if (!isCompoundProperty(name))
+    throw std::runtime_error("Algorithm::setIndexProperty can only be used "
+                             "with properties declared using "
+                             "declareIndexProperty.");
+
+  auto *wsProp =
+      dynamic_cast<WorkspaceProperty<WsType> *>(getPointerToProperty(name));
+  auto *indexTypeProp = dynamic_cast<IndexTypeProperty *>(
+      getPointerToProperty(name + "IndexType"));
+  auto *indexProp =
+      dynamic_cast<IndexProperty *>(getPointerToProperty(name + "IndexSet"));
+
+  setWorkspaceProperty<WsType, T1>(
+      wsProp, wksp, boost::is_convertible<T1, boost::shared_ptr<WsType>>());
+
+  *indexTypeProp = type;
+
+  *indexProp = list;
+}
+
+/** Mechanism for setting the index property with a workspace shared pointer.
+* This method can only be used if T1 is convertible to a MatrixWorkspace and
+* T2 is either std::string or std::vector<int>
+
+@param name Property name
+@param wksp Workspace as a pointer
+@param type Index type IndexType::WorkspaceIndex or IndexType::SpectrumNum
+@param list List of indices to be used.
+*/
+template <typename T1, typename T2, typename, typename>
+void Algorithm::setIndexProperty(const std::string &name,
+                                 const boost::shared_ptr<T1> &wksp,
+                                 IndexType type, const T2 &list) {
+  doSetIndexProperty<boost::shared_ptr<T1>, T2, T1>(name, wksp, type, list);
+}
+
+/** Mechanism for setting the index property with a workspace shared pointer.
+* This method can only be used if T1 is convertible to a MatrixWorkspace and
+* T2 is either std::string or std::vector<int>
+
+@param name Property name
+@param wsName Workspace as a pointer
+@param type Index type IndexType::WorkspaceIndex or IndexType::SpectrumNum
+@param list List of indices to be used.
+*/
+template <typename T1, typename T2, typename, typename>
+void Algorithm::setIndexProperty(const std::string &name,
```

Comment:
Because `WorkspaceProperty` supports this as users, especially in the context of python scripts, will expect to be able to set the workspace using its name.